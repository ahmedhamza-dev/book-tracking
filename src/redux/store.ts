import { configureStore } from "@reduxjs/toolkit";
import bookSlice from "./slices";

const store = configureStore({
    reducer: {
        allBooks: bookSlice.reducer
    }
});

export default store;