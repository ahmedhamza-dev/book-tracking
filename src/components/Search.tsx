import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { bookSearch } from "../redux/slices";
import SearchShelf from "./SearchShelf";

const Search = () => {
  const [search, setSearch] = useState("");
  const state: any = useSelector((state) => state);
  const dispatch = useDispatch<any>();
  const { allBooks } = state;

  const searchHanlding = async (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(event.target.value);
  };
  useEffect(
    () => {
      dispatch(bookSearch(search));
    },
    [search]
  );
  return (
    <div className="search-books">
      <div className="search-books-bar">
        <Link to="/" className="close-search">
          Close
        </Link>
        <div className="search-books-input-wrapper">
          <input
            type="text"
            placeholder="Search by title or author"
            onChange={searchHanlding}
            data-testid="search-input"
          />
        </div>
      </div>
      <div className="search-books-results">
        <ol className="books-grid" data-testid="books-grid">
          {allBooks.booksFromSearch && (
            <SearchShelf
              books={allBooks.booksFromSearch}
            />
          )}
        </ol>
      </div>
    </div>
  );
};

export default Search;
