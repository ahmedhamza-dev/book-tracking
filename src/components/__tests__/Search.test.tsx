import React from "react";
import "@testing-library/jest-dom";
import Search from "../Search";
import renderer from "react-test-renderer";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { BrowserRouter as Router, Switch } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../../redux/store";

afterEach(() => {
  cleanup();
});

it("Search component test", () => {
  const searchComponent = (
    <>
      <Provider store={store}>
        <Router>
          <Switch>
            <Search />
          </Switch>
        </Router>
      </Provider>
    </>
  );

  const component = renderer.create(searchComponent);
  const { getByTestId, getByText } = render(searchComponent);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const link = getByText("Close");
  fireEvent.click(link);
  expect(window.location.pathname).toBe("/");

  const searchInput: any = getByTestId("search-input");
  expect(searchInput).toBeInTheDocument();
  fireEvent.change(searchInput, {
    target: { value: "Robitics" },
  });
  expect(searchInput.value).toEqual("Robitics");

  const booksGrid = getByTestId("books-grid");
  expect(booksGrid).toBeInTheDocument();
});
