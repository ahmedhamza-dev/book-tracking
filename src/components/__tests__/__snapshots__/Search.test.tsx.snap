// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component test 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      onClick={[Function]}
    >
      Close
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        data-testid="search-input"
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
      data-testid="books-grid"
    />
  </div>
</div>
`;
